python/lsst/daf/butler/sql/queryBuilder.py

```diff
@@ -113,10 +121,57 @@ def visitBinaryOp(self, operator, lhs, rhs, node):
 
     def visitIsIn(self, lhs, values, not_in, node):
         # Docstring inherited from TreeVisitor.visitIsIn
+
+        # `values` is a list of literals and ranges, range is represented
+        # by a tuple (start, stop, stride). We need to transform range into
+        # some SQL construct, simplest would be to generate a set of literals
+        # and add it to the same list but it could become too long. What we
+        # do here is to introduce some large limit on the total number of
+        # items in IN() and if range exceeds that limit then we do something
+        # like:
+        #
+        #    X IN (1, 2, 3)
+        #    OR
+        #    (X BETWEEN START AND STOP AND MOD(X, STRIDE) = MOD(START, STRIDE))
+        #
+        # or for NOT IN case
+        #
+        #    X NOT IN (1, 2, 3)
+        #    AND NOT
+        #    (X BETWEEN START AND STOP AND MOD(X, STRIDE) = MOD(START, STRIDE))
+
+        max_in_items = 1000
+
+        # split the list into literals and ranges
+        literals, ranges = [], []
+        for item in values:
+            if isinstance(item, tuple):
+                ranges.append(item)
+            else:
+                literals.append(item)
+
+        ranges_remain = []
+        for start, stop, stride in ranges:
+            count = (stop - start + 1) // stride
+            if len(literals) + count > max_in_items:
+                ranges_remain.append((start, stop, stride))
+            else:
+                # add all values to literal list, stop is inclusive
+                literals += [literal(value) for value in range(start, stop+1, stride)]
+
+        clauses = [lhs.in_(literals)]
+        for start, stop, stride in ranges_remain:
+            # X BETWEEN START AND STOP AND MOD(X, STRIDE) = MOD(START, STRIDE)
+            expr = lhs.between(start, stop)
+            if stride != 1:
+                expr = and_(expr, (lhs % stride) == (start % stride))
+            clauses.append(expr)
+
         if not_in:
-            return lhs.notin_(values)
+            clauses = [not_(expr) for expr in clauses]
```

Comment:
it seems a shame to need to have a loop over the len(ranges_remain) twice in the case of not_in just to add the not expression. Consider putting the if not_ also in the above loop, something like:
`clauses.append(not_(expr) if not_in else expr)`
Or of course the converse Boolean logic

Reply:
I replaced it with

        expr = or_(*clauses)
        if not_in:
            expr = not_(expr)

it make slightly different expression structure but it should be logically equivalent indeed.

Reply:
I'm not sure I understand the change you made. The change I was thinking of was having two if not_in blocks, one that is the returns, and the other inside where you are creating the expr for the first time. I think that would produce exactly the same thing you had, but with one less loop over all the strides.

Reply:
Oh, I see, you do not like two loops there? I guess this is how I first thought of it and it ended up in implementation. I don't see big issue with iterating twice over ranges, should not be any overhead compared to constructing all expressions. Still I can merge them into this if you prefer:

        literals, ranges = [], []
        for item in values:
            if isinstance(item, tuple):
                ranges.append(item)
            else:
                literals.append(item)

        clauses = []
        for start, stop, stride in ranges:
            count = (stop - start + 1) // stride
            if len(literals) + count > max_in_items:
                # X BETWEEN START AND STOP AND MOD(X, STRIDE) = MOD(START, STRIDE)
                expr = lhs.between(start, stop)
                if stride != 1:
                    expr = and_(expr, (lhs % stride) == (start % stride))
                clauses.append(expr)
            else:
                # add all values to literal list, stop is inclusive
                literals += [literal(value) for value in range(start, stop+1, stride)]
        
        if literals:
            # add IN() in front of BETWEENs
            clauses.insert(0, lhs.in_(literals))

        expr = or_(*clauses)
        if not_in:
            expr = not_(expr)

        return expr

I prefer to handle `not_in` in one place, makes it clearer what is going on (at least for myself)